### PR TITLE
Cleanup/refactor sync code

### LINKED ISSUE #418: pushChanges is missing a case
The `then` block in https://github.com/Kinto/kinto.js/blob/d696e76dfba7f36a107f0d3a910cee142bbd39d7/src/collection.js#L854-L873 has a set of if-else-if blocks that aren't exhaustive. If we get something `resolved` which is not synced (`resolvedUnsynced.length != 0`), and our strategy isn't either `CLIENT_WINS` or `SERVER_WINS`, we fail to return anything at all. This caused me problems when I was trying to put together a fix for #414 and tried to mark a conflict as "resolved".

I'm not sure if we consider this case something that "can't happen". It seems like the only strategy this applies to is `MANUAL`, and we could assume that unsynced resolutions should never get to `pushChanges` in `MANUAL` because the conflicts stopped the sync. If this is how we want to hande this, I think we should put a default "else" case in `pushChanges` that throws a "can't happen" error, or at the very least add a comment explaining how someone got here and suggesting what they do to fix it.

### PR DESCRIPTION
**Pull request by leplatrem**
Now that we have switched to async/await it's a lot easier to read the code of the sync process and spot inconsistencies :)

I did that refactoring by keeping an eye on the integration tests. I'll have to change the specs/code of the unit tests now.

* [x] Refactor / added some comments
* [x] Update docstrings
* [ ] Document changes
* [x] Make unit tests pass

Fixes #418 #340

### CODE REVIEW

File: src/collection.js [hunk 1]
```diff
@@ -884,59 +950,24 @@ export default class Collection {
...
-        return {data: synced};
-      });
-      const deleted = toDeleteLocally.map((record) => {
-        transaction.delete(record.id);
-        // Amend result data with the deleted attribute set
-        return {data: {id: record.id, deleted: true}};
-      });
-      return updated.concat(deleted);
-    });
+    await this.importPublications(syncResultObject, decoded, options.strategy);
```
Comment (leplatrem) [inline src/collection.js]: note: add `if decoded.length > 0`

**Comment by leplatrem** [comment]
@n1k0 @glasserc r? :)

**Comment by leplatrem** [comment]
As a summary, here is what I did:
* `pushChanges()` is not in charge of `gatheringChanges()` anymore
* `importChanges()` is not in charge of decoding the records anymore
* created an `importPublications()` for `pushChanges()` results to mimic the code of `pullChanges()` with `importChanges()`
* I simplified the way `_handleConflicts()` is called, and changed it so that shares the same transaction as `importChanges()` and `importPublications()`
* the lastmodified option is now used explicitly in the second `pullChanges()` during sync instead of storing intermediate values as member variable during the sync process.

File: src/collection.js [hunk 2]
```diff
@@ -1047,33 +1045,40 @@ export default class Collection {
...
     try {
-      const result = new SyncResultObject();
-      const pullResult = await this.pullChanges(client, result, options);
-      const pushResult = await this.pushChanges(client, pullResult, options);
-      // Avoid performing a last pull if nothing has been published.
-      if (pushResult.published.length === 0) {
-        syncResult = pushResult;
-      } else {
+      // Fetch last changes from the server.
+      await this.pullChanges(client, result, options);
```
Comment (glasserc) [inline src/collection.js]: @leplatrem explained to me that although we used to have this sort of return-value chaining, which suggests immutability, things were also being modified in-place, so he took out the "fake" return values.

File: src/collection.js [hunk 3]
```diff
@@ -652,33 +660,65 @@ export default class Collection {
...
+      return {published, resolved};
+    });
+
+    syncResultObject.add("published", published);
+
+    if (resolved.length > 0) {
+      // Resolved conflicts are merged with previous ones.
+      const resolvedIds = resolved.map((r) => r.id);
+      const mergedResolved = syncResultObject.resolved.filter((r) => resolvedIds.indexOf(r.id) < 0)
+                                                      .concat(resolved);
```
Comment (glasserc) [inline src/collection.js]: Can we do better here? Maybe we should have a `Set` of resolved records, or at least a `Set` of resolved IDs. The indexOf call seems a little ugly to me. Maybe this complexity ought to be in the `SyncResultObject#add` method?

File: src/collection.js [hunk 4]
```diff
@@ -667,6 +656,73 @@ export default class Collection {
...
+        return synced;
+      });
+      const deleted = toDeleteLocally.map((record) => {
+        transaction.delete(record.id);
+        // Amend result data with the deleted attribute set
+        return {id: record.id, deleted: true};
+      });
+      const published = updated.concat(deleted);
+      // Handle conflicts, if any
+      const resolved = this._handleConflicts(transaction, conflicts, strategy);
```
Comment (glasserc) [inline src/collection.js]: It seems a little strange to have `importPublications`, which I assumed was just going to handle updating state about records that were successfully published, to also handle conflicts (records that weren't actually published). Can we have the caller manage the transaction and call both `importPublications` and `_handleConflicts`? Or maybe we can change the function name to something like `importPushResult`, and make `conflicts` be a parameter to the function?

File: src/collection.js [hunk 5]
```diff
@@ -667,6 +656,73 @@ export default class Collection {
...
   /**
+   * Imports the responses of pushed changes into the local database.
+   * Basically it stores the timestamp assigned by the server into the local
+   * database.
+   * @param  {SyncResultObject} syncResultObject The sync result object.
+   * @param  {Array}            toApplyLocally   The list of changes to import in the local database.
+   * @param  {String}           strategy         The {@link Collection.strategy}.
+   * @return {Promise}
+   */
+  async importPublications(syncResultObject, toApplyLocally, strategy) {
```
Comment (n1k0) [inline src/collection.js]: Same remark wrt default strategy value

File: src/collection.js [hunk 6]
```diff
@@ -667,6 +656,73 @@ export default class Collection {
...
+    const toDeleteLocally = toApplyLocally.filter((r) => r.deleted);
+    const toUpdateLocally = toApplyLocally.filter((r) => !r.deleted);
+
+    const conflicts = syncResultObject.conflicts;
```
Comment (n1k0) [inline src/collection.js]:
```js
const {conflicts} = syncResultObject; 
```

File: src/collection.js [hunk 7]
```diff
@@ -768,8 +825,12 @@ export default class Collection {
     if (!syncResultObject.ok) {
       return syncResultObject;
     }
+
+    const since = this.lastModified ? this.lastModified
+                                    : await this.db.getLastModified();
+
     options = {strategy: Collection.strategy.MANUAL,
```
Comment (n1k0) [inline src/collection.js]: Nit: while you're here, please format this object like everywhere else

File: src/collection.js [hunk 8]
```diff
@@ -837,18 +907,18 @@ export default class Collection {
...
    * @param  {KintoClient.Collection} client           Kinto client Collection instance.
    * @param  {SyncResultObject}       syncResultObject The sync result object.
+   * @param  {Object}                 changes          The change object.
+   * @param  {Array}                  changes.toDelete The list of records to delete.
+   * @param  {Array}                  changes.toSync   The list of records to create/update.
    * @param  {Object}                 options          The options object.
    * @return {Promise}
    */
-  async pushChanges(client, syncResultObject, options={}) {
+  async pushChanges(client, {toDelete=[], toSync}, syncResultObject, options={}) {
```
Comment (n1k0) [inline src/collection.js]: Now I'm wondering if this new second argument shouldn't be split as two distinct ones. Thoughts?

File: src/collection.js [hunk 9]
```diff
@@ -1047,33 +1057,43 @@ export default class Collection {
...
-      const result = new SyncResultObject();
-      const pullResult = await this.pullChanges(client, result, options);
-      const pushResult = await this.pushChanges(client, pullResult, options);
-      // Avoid performing a last pull if nothing has been published.
-      if (pushResult.published.length === 0) {
-        syncResult = pushResult;
-      } else {
+      // Fetch last changes from the server.
+      await this.pullChanges(client, result, options);
+      const {lastModified} = result;
```
Comment (n1k0) [inline src/collection.js]: Nit
```js
const {ok, lastModified, resolved, published} = result;
```

So you can directly reference these later.

File: src/collection.js [hunk 10]
```diff
@@ -1047,33 +1064,43 @@ export default class Collection {
         new Error(`Server is asking clients to back off; retry in ${seconds}s or use the ignoreBackoff option.`));
     }
 
-    this._lastModified = await this.db.getLastModified();
```
Comment (glasserc) [inline src/collection.js]: Why are we getting rid of this?

**Pull request merged**